src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterDocumentRewriter.cs

```diff
@@ -0,0 +1,693 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.IntroduceVariable
+{
+    internal abstract partial class AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax>
+    {
+        private class IntroduceParameterDocumentRewriter
+        {
+            private readonly AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax> _service;
+            private readonly Document _originalDocument;
+            private readonly SyntaxGenerator _generator;
+            private readonly ISyntaxFactsService _syntaxFacts;
+            private readonly ISemanticFactsService _semanticFacts;
+            private readonly TExpressionSyntax _expression;
+            private readonly IMethodSymbol _methodSymbol;
+            private readonly SyntaxNode _containerMethod;
+            private readonly IntroduceParameterCodeActionKind _actionKind;
+            private readonly bool _allOccurrences;
+
+            public IntroduceParameterDocumentRewriter(AbstractIntroduceParameterService<TExpressionSyntax, TInvocationExpressionSyntax, TObjectCreationExpressionSyntax, TIdentifierNameSyntax> service,
+                Document originalDocument,
+                SyntaxGenerator generator,
+                ISyntaxFactsService syntaxFacts,
+                ISemanticFactsService semanticFacts,
```

Comment:
note: you can make it so that these are not passed in here.  instead, you can derive it from `originalDocument`.  up to you.

Reply:
yeah that is cleaner